src/dodal/devices/tetramm.py

```diff
@@ -0,0 +1,273 @@
+        are not missed.
+
+        Args:
+            frame_time (float): The time for a single frame in seconds
+
+        Raises:
+            ValueError: If frame_time is too low to collect the required number
+            of readings per frame.
+        """
+
+        values_per_reading: int = int(
+            frame_time * self.base_sample_rate / self.readings_per_frame
+        )
+
+        if values_per_reading < self.minimum_values_per_reading:
+            raise ValueError(
+                f"frame_time {frame_time} is too low to collect at least "
+                f"{self.minimum_values_per_reading} values per reading, at "
+                f"{self.readings_per_frame} readings per frame."
+            )
+        await self._drv.values_per_reading.set(values_per_reading)
+
+    @property
+    def max_frame_rate(self) -> float:
+        """Max frame rate in Hz for the current configuration"""
+        return 1 / self.minimum_frame_time
+
+    @max_frame_rate.setter
+    def max_frame_rate(self, mfr: float):
+        self.minimum_frame_time = 1 / mfr
+
+    @property
+    def minimum_frame_time(self) -> float:
+        time_per_reading = self.minimum_values_per_reading / self.base_sample_rate
+        return self.readings_per_frame * time_per_reading
+
+    @minimum_frame_time.setter
+    def minimum_frame_time(self, frame: float):
+        time_per_reading = self.minimum_values_per_reading / self.base_sample_rate
+        self.readings_per_frame = int(
+            min(self.maximum_readings_per_frame, frame / time_per_reading)
+        )
+
+
+IDLE_TETRAMM = {
+    "drv.acquire": 0,
+}
+
+COMMON_TETRAMM = {
+    "drv.range": TetrammRange.uA,
+    "drv.num_channels": "4",
+    "drv.resolution": TetrammResolution.TwentyFourBits,
+    "drv.geometry": TetrammGeometry.Square,
+}
+
+TRIGGERED_TETRAMM = {
+    **COMMON_TETRAMM,
+    "drv.trigger_mode": TetrammTrigger.ExtTrigger,
+}
+
+FREE_TETRAMM = {
+    **COMMON_TETRAMM,
+    "drv.values_per_reading": 10,
+    "drv.averaging_time": 0.1,
+    "drv.trigger_mode": TetrammTrigger.FreeRun,
+}
+
+
+def triggered_tetramm(dev: TetrammDriver) -> Generator[Msg, None, None]:
+    sigs = walk_rw_signals(dev)
+    yield from set_signal_values(
+        sigs,
+        [
+            IDLE_TETRAMM,
+            TRIGGERED_TETRAMM,
+        ],
+    )
+
+
+def free_tetramm(dev: TetrammDriver) -> Generator[Msg, None, None]:
+    """Freerun the tetramm, setting it to acquire so diodes update."""
+    sigs = walk_rw_signals(dev)
+    yield from set_signal_values(sigs, [IDLE_TETRAMM, FREE_TETRAMM])
+    yield from bps.abs_set(dev.drv.acquire, 1)
+
+
+class TetrammShapeProvider(ShapeProvider):
+    max_channels = 11
+
+    def __init__(self, controller: TetrammController) -> None:
+        self.controller = controller
+
+    async def __call__(self) -> Sequence[int]:
+        return [self.max_channels, self.controller.readings_per_frame]
```

Comment:
Can we use the value of `drv.num_channels`? That seems to be the actual shape when deployed on p38

Reply:
So the actual shape should be `[drv.num_channels, controller.readings_per_frame]`?

Reply:
That is my understanding. On p38 I have /entry/data/data with actual shape [1, 1000], if it is related to the number of channels, it shouldn't be the max_channels (wherever that comes from) but the configured number of channels

Reply:
...In conclusion, we're not sure, @Tom-Willemsen or @tpoliaw, do you have any input?

Reply:
From a purely theoretical perspective the currently-configured number of channels makes more sense to me than the maximum number of channels... but I don't really know much about this specifically.

What is currently done in GDA?

Reply:
In GDA NcdTetrammDetector ~ `controller.dimensions = [*scan_shape, frameCount]`, but `getChannels()` is used as the value of `getDataDimensions()`

Reply:
(i.e., shapeProvider should only [frameCount] for the former, but [channels, frameCount] if the latter

Reply:
I wouldn't use p38 as an example of data shape, I don't think it is correctly set up to record all the channels whatever the value of `num_channels`. From memory, I think the data written out on I22 is always the same size as well, just some of the channels are 0 if the num channels is set below the max. **Please double check this**, it's been a while since I've worked with it.

Reply:
Thanks, I'm going to suggest that we keep this as it was in November (i.e. as it is in this PR now) and raise additional tickets based on Tuesday testing if need be. What does everyone think?

Reply:
I'd agree, presumably it was not-wrong for November and we can investigate further when we need another TetrAMM: if the p38 one is configured differently we should try and get it configured the same, it has been updated to use the same IOC version as the one on i22 IIRC.